src/regularize/diffs.jl

```diff
@@ -1,99 +1,122 @@
 without using `spdiagm` (or any `SparseArrays` function).
 
 in
-- `X`		`M x N` array (typically a 2D image).
-It cannot be a Vector!  (But it can be a `Mx1` or `1xN` 2D array.)
+- `X` `N_1 × ... × N_d` array (typically an N-D image).
+
+option
+- `dims` dimensions along which to perform finite differences; default 1:d
 
 out
-- `d`		vector of length `N*(M-1) + (N-1)*M`
+- `d` vector of length `N_d*...*(N_1-1) + ... + (N_d-1)*...*N_1`
 """
-function diff2d_forw(x::AbstractMatrix{<:Number})
-	return [diff(x,dims=1)[:]; diff(x,dims=2)[:]]
+function diffnd_forw(x::AbstractArray{<:Number,D} ; dims=1:D) where {D}
+    return reduce(vcat, vec(diff(x, dims = d)) for d in dims)
 end
 
 
 """
-`z =  diff2d_adj(d, M, N; out2d=false)`
+    z =  diffnd_adj(d, N... ; dims=1:length(N), out2d=false)
 
-Adjoint of 2D finite differences along both dimensions.
+Adjoint of N-D finite differences along both dimensions.
 Performs the same operations as
-``z = [(I_N \\otimes D_M); (D_N \\otimes I_M)]' * d``
-where D_N denotes the N-1 x N 1D finite difference matrix
-and \\otimes denotes the Kronecker product,
-but does it efficiently without using spdiagm (or any SparseArrays function).
+``z = [(I_{N_d} \\otimes \\cdots \\otimes D_{N_1}); \\dots; (D_{N_d} \\otimes \\cdots \\otimes I_{N_1})]' * d``
+where D_N denotes the `N-1 × N` 1D finite difference matrix
+and `\\otimes` denotes the Kronecker product,
+but does it efficiently without using `spdiagm` (or any `SparseArrays` function).
 
 in
-- `d`		vector of length `N*(M-1) + (N-1)*M`
-- `M,N`		desired output size
+- `d` vector of length `N_d*...*(N_1-1) + ... + (N_d-1)*...*N_1`
+- `N...` desired output size
 
 option
-- `out2d`	if true then return `M x N` array, else `M*N` vector
+- `dims` dimensions along which to perform adjoint finite differences
+- `outnd` if true then return `N_1 × ... × N_d` array, else `prod(N)` vector
 
 out
-- `z`		`M*N` vector or `M x N` array (typically a 2D image)
+- `z` `prod(N)` vector or `N_1 × ... × N_d` array (typically an N-D image)
 
 """
-function diff2d_adj(d::AbstractVector{<:Number}, M::Int, N::Int ; out2d=false)
-
-	length(d) != N*(M-1) + (N-1)*M && throw("length(d)")
-
-	dx = d[1:(N*(M-1))]
-	dx = reshape(dx, M-1, N)
-	dy = d[1+(N*(M-1)):end]
-	dy = reshape(dy, M, N-1)
-	zx = [-transpose(dx[1,:]);
-		(@views dx[1:end-1,:] - dx[2:end,:]);
-		transpose(dx[end,:])]
-	zy = [-dy[:,1] (@views dy[:,1:end-1] - dy[:,2:end]) dy[:,end]]
-	z = zx + zy # M by N
-
-	return out2d ? z : z[:] # array or vector?
+function diffnd_adj(d::AbstractVector{<:Number}, N::Vararg{Int,D}
+    ; dims=1:D, outnd=false) where {D}
+
+# todo: diff2d_adj etc. 1-liners for backward compat
+# todo: N::Dims instead of Vararg?
+# todo: remove outnd option for type inference?
+# todo: discuss, >1 should be needed only for the "dims" dimensions?
+    # Note that N must be strictly greater than 1 for each dimension specified
+    # by dims, or N must be 1 for all dimensions specified by dims
```

Comment:
I wrote the original comment before adding `dims`. So, yes, `N > 1` is only necessary for `N[dims]`.